File: leetcode/Arrays_Hash/Top K Frequent Elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> count;
        for (int num : nums) count[num]++;
        
        vector<vector<int>> buckets(nums.size() + 1);

        for (auto pair : count){
            int number = pair.first;
            int freq = pair.second;
            buckets[freq].push_back(number);
        }
        vector<int> res;

        for (int i = buckets.size() - 1; i >= 0 && res.size() < k; i--){
            for (int num : buckets[i]){
                res.push_back(num);
                if (res.size() >= k) return res;
            }
        }
        return res;
        
    }
};

File: leetcode/Arrays_Hash/Top K Frequent Elements.py
import heapq
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        # Step 1: Count frequency
        count = Counter(nums)
        print(count)
        
        # Step 2: Create buckets
        # Max possible frequency = len(nums)
        buckets = [[] for _ in range(len(nums) + 1)]
        print(buckets)
        
        for num, freq in count.items():
            buckets[freq].append(num)
        print(buckets)
        
        # Step 3: Collect top k frequent
        result = []
        for freq in range(len(buckets) - 1, 0, -1):
            for num in buckets[freq]:
                result.append(num)
                if len(result) == k:
                    return result